### PR TITLE
Expand failure_mocks: helpers, constants, tests, and docs

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds centralized failure message constants and new helpers for failure-mock testing
- Introduces `setup_failure_mocks` and `assert_failure_error` to drive and verify mocked failures across toolchain channels
- Adds `failure_summary` to describe failure scenarios and uses exact-message matching for failure paths
- Refactors tests to rely on new helpers and constants; updates tests to ensure exact messages
- Updates documentation to reflect new helpers and usage

### Testing Utilities
- Added `TOOLCHAIN_INSTALL_FAILURE_MESSAGE` and `COMPONENT_INSTALL_FAILURE_MESSAGE` constants for standardized failure messaging
- Added `setup_failure_mocks(runner, seq, channel, setup)` to configure mocks for a given toolchain setup according to a `setup` variant. It wires the mock command sequence so that `ensure_installed_with` will trigger the corresponding error variant. Any extra components in `setup.additional_components` are included in the expected `rustup component add` invocation where applicable
- Added `assert_failure_error(err, channel, setup)` to validate that a given `InstallerError` matches the expected variant and fields for the specified `setup.failure` on the toolchain `channel`. The function panics with descriptive messages if the variant or fields do not match. For `ComponentAdd`, it also verifies that the reported component list includes the required components plus any in `setup.additional_components`

### Failure Matching Helpers
- Added `failure_summary(setup)` to produce a human-readable summary for the given `setup`'s failure
- Added `assert_failure_error(err, channel, setup)` to assert that `err` matches the expected failure variant/fields; uses `failure_summary` for messages and is `is_component_install_failed` for the `ComponentAdd` path

### Guard Logic
- Keeps the pre-guard that returns false when the toolchain does not match, and simplifies the core condition to rely on component equality and the exact failure message constant (no redundant toolchain check in the core condition)

### Documentation
- Expanded doc comments to explain the purpose, usage, and edge cases for the new helpers, ensuring future contributors understand their behavior and expectations

### Tests
- Updated installer tests to reference the new failure message constants
- Added unit tests to verify the failure matching helpers behavior (e.g., exact message matching for component failures, failure_summary formatting)
- Updated expectations in existing tests to rely on `COMPONENT_INSTALL_FAILURE_MESSAGE` instead of the raw string

## Test plan
- Run `cargo test` for the installer tests to verify that the new helpers integrate correctly with existing failure mocks
- Ensure that existing tests relying on failure mocks continue to pass
- If needed, add or adjust unit tests that exercise `setup_failure_mocks` and `assert_failure_error` directly to cover common failure scenarios

---
◳ Generated by [DevBoxer](https://www.devboxer.com)

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/d893ca4c-caff-43bf-bde5-308bd2d1fa10

📎 **Task**: https://www.devboxer.com/task/3af976c2-4ae8-40dc-9702-a8f52931adee

📎 **Task**: https://www.devboxer.com/task/8651f091-767b-46e2-958f-95a77231bff5

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- In `is_component_install_failed`, relying on `message.contains("component failed")` is fairly brittle; consider matching on a more structured error indicator (e.g., a specific prefix, code, or pattern) so small wording changes in the error message don’t silently break the guard.
- For `assert_failure_error`, consider including both the `channel` and a summary of `setup.failure` in the panic messages to make debugging failures easier when multiple toolchains or failure variants are exercised in the same test run.

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | New test helper utilities for installer toolchain testing lack documentation in developers' guide and roadmap tracking. | Document the failure mock helpers in docs/developers-guide.md with usage examples and update docs/roadmap.md accordingly. |
| Module-Level Documentation | ⚠️ Warning | The failure\_mocks.rs module lacks documentation for the `as_str()` method and two public constants, though other public functions are properly documented. | Add `///` documentation comments to the `as_str()` method, `TOOLCHAIN_INSTALL_FAILURE_MESSAGE`, and `COMPONENT_INSTALL_FAILURE_MESSAGE` constant explaining their purpose and usage. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?


#### installer/src/toolchain/tests/mod.rs (1)

`169-189`: _🧹 Nitpick_ | _🔵 Trivial_
**Eliminate literal failure-marker drift in test setup.**
Replace the hard-coded stderr marker with the imported constant in the same test block, so setup and assertion cannot diverge.

##### Diff

```diff
-    let runner = CapturingCommandRunner::new(output_with_stderr(1, "component failed"));
+    let runner = CapturingCommandRunner::new(output_with_stderr(
+        1,
+        COMPONENT_INSTALL_FAILURE_MESSAGE,
+    ));
```

##### 🤖 Prompt for AI Agents

File: crates/no_expect_outside_tests/src/lib_ui_tests.rs [hunk 1]
```diff
@@ -287,15 +287,28 @@ fn is_dependency_rlib(path: &Path, prefix: &str) -> bool {
...
+    ));
+}
+
+#[test]
+fn rstest_example_compiles_under_test_harness() {
+    run_example_under_test_harness(&ExampleHarnessRun::new(
+        "pass_expect_in_rstest_harness",
+        "rstest",
+    ));
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/no_expect_outside_tests/src/lib_ui_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?


#### installer/src/toolchain/tests/mod.rs (1)

`169-189`: _🧹 Nitpick_ | _🔵 Trivial_
**Eliminate literal failure-marker drift in test setup.**
Replace the hard-coded stderr marker with the imported constant in the same test block, so setup and assertion cannot diverge.

##### Diff

```diff
-    let runner = CapturingCommandRunner::new(output_with_stderr(1, "component failed"));
+    let runner = CapturingCommandRunner::new(output_with_stderr(
+        1,
+        COMPONENT_INSTALL_FAILURE_MESSAGE,
+    ));
```

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | New public functions and constants lack comprehensive test coverage for panic paths, error handling, and private helper functions. | Add #\[should\_panic\] unit tests for assert\_failure\_error() panic conditions, verify panic message format includes channel and failure\_summary, and test private setup functions in isolation. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (1 error, 1 warning, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | PR introduces public testing helpers but test coverage is insufficient; core validation logic lacks unit tests verifying correct inputs pass without panic. | Add unit tests calling assert\_failure\_error() with correctly-matching error inputs for all three InstallFailure variants, and tests verifying panic on field mismatches. |
| Testing (Unit) | ⚠️ Warning | PR includes tests that exercise private test infrastructure and helpers rather than meaningful application behaviour or critical invariants. | Remove tests exercising only private helpers: failure\_summary\_lists\_requested\_components, setup\_component\_add\_failure\_mocks\_inner\_drives\_component\_failure\_in\_isolation, and assert\_failure\_error\_panics\_with\_various\_channels\_and\_summaries. Retain component\_failure\_match\_rejects\_superstring\_messages and rely on ensure\_installed\_reports\_failure for functional coverage. |
| Module-Level Documentation | ❓ Inconclusive | Repository does not contain the files referenced in PR summary, preventing verification of module-level documentation in failure\_mocks.rs. | Ensure repository contains all referenced files and provide access to installer/src/toolchain/tests/failure\_mocks.rs for documentation inspection. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational.

### ❌ Failed checks (1 error, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Integration test only validates the happy path where mocks emit correct values; field equality checks could be removed without affecting test passage, leaving no coverage for field mismatches. | Add explicit negative test cases constructing errors with correct variants but incorrect field values, asserting `assert_failure_error` panics appropriately. |
| Developer Documentation | ⚠️ Warning | The PR introduces undocumented runtime shell resolution for CARGO and MDLINT variables in the Makefile, lacking explanation of fallback logic and path preferences in developers-guide.md. | Document the $\left(\right. o r$ (shell...)) fallback pattern, specify path preference order, and explain why runtime resolution improves development experience over fixed command names. |

**Comment by leynos** [comment]
@coderabbitai Please investigate the cause of the following issue, identify a fix and provide an AI coding agent prompt for the fix:

~~~
Run make lint
  make lint
  shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
  env:
    CARGO_TERM_COLOR: always
    CS_ACCESS_TOKEN: 
    CODESCENE_CLI_SHA256: 
    RUSTFLAGS: -D warnings
    RUSTDOCFLAGS: -D warnings
    CARGO_INCREMENTAL: 0
    CARGO_PROFILE_DEV_DEBUG: 0
    RUST_BACKTRACE: short
    CARGO_UNSTABLE_SPARSE_REGISTRY: true
    CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
    UV_TOOL_BIN_DIR: D:\a\_temp\uv-tool-bin-dir
    UV_TOOL_DIR: D:\a\_temp\uv-tool-dir
    UV_CACHE_DIR: D:\a\_temp\setup-uv-cache
    SCCACHE_PATH: C:\hostedtoolcache\windows\sccache\0.14.0\x64/sccache
    ACTIONS_CACHE_SERVICE_V2: on
    ACTIONS_RESULTS_URL: https://results-receiver.actions.githubusercontent.com/
    ACTIONS_RUNTIME_TOKEN: ***
    UV_PYTHON_INSTALL_DIR: D:\a\_temp\uv-python-dir
RUSTDOCFLAGS="--cfg docsrs -D warnings" C:\Users\runneradmin/.cargo/bin/cargo doc --workspace --no-deps
/usr/bin/sh: line 1: C:Usersrunneradmin/.cargo/bin/cargo: No such file or directory
make: *** [Makefile:101: lint] Error 127
Error: Process completed with exit code 2.
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue.

### ❌ Failed checks (3 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | File exceeds 400-line limit (471 lines). Review explicitly requested test module extraction to separate file, which remains unaddressed. | Extract test module (lines 250–471) into installer/src/toolchain/tests/failure\_mocks/tests.rs to comply with repository line-limit policy. |
| Module-Level Documentation | ❌ Error | Two public functions lack required `# Arguments` sections: `setup_failure_mocks()` and `assert_failure_error()` have undocumented parameters violating Rust documentation standards. | Add `# Arguments` sections to both functions documenting all parameters, and add `# Panics` sections where applicable per Rust rustdoc standards. |
| Unit Architecture | ❌ Error | File `installer/src/toolchain/tests/failure_mocks.rs` exceeds the 400-line hard constraint at 471 lines, and the 221-line test module remains nested rather than extracted to a sibling module as explicitly requested in review feedback. | Extract the `#[cfg(test)] mod tests` block into `installer/src/toolchain/tests/failure_mocks/tests.rs`, update the parent file to declare `#[cfg(test)] mod tests;` alone, and adjust visibility annotations as needed to reduce the primary file to approximately 250 lines. |
| Testing (Unit) | ⚠️ Warning | Unit tests in failure\_mocks.rs exercise private helper functions directly rather than verifying meaningful functional behaviour through observable system boundaries. | Remove the unit test module from failure\_mocks.rs (lines 250–471) and extract remaining helper code into a separate file below the 400-line limit. |
